lib/bridge/QuitDebouncer.js

```diff
@@ -0,0 +1,125 @@
+    // This is to smooth incoming PART spam from IRC clients that suffer from a
+    // net-split (or other issues that lead to mass PART-ings)
+    this._debouncerForServer = {
+      // $server.domain: {
+            // $nick: {
+            //   promise: Promise, //Promise that resolves if the user joins a channel having quit
+            //   resolve: Function //function to call when a user joins
+            // }
+      // }
+    };
+
+    // Keep a track of the times at which debounceQuit was called, and use this to
+    // determine the rate at which quits are being received. This can then be used
+    // to detect net splits.
+    /*$server : [1477386173850, 1477386173825, ...]*/
+    this._quitTimesMs = {};
+    Object.keys(this.ircBridge.config.ircService.servers).forEach((domain) => {
+        this._quitTimesMs[domain] = [];
+    });
+}
+
+/**
+ * Called when the IrcHandler receives a JOIN. This resolves any promises to join that were made
+ * when a quit was debounced during a split.
+ * @param {string} nick The nick of the IRC user joining.
+ * @param {IrcServer} server The sending IRC server.
+ */
+QuitDebouncer.prototype.onJoin = function (nick, server) {
+    if (!this._debouncerForServer[server.domain]) {
+        return;
+    }
+    let rejoin = this._debouncerForServer[server.domain][nick];
+    if (rejoin) {
+        rejoin.resolve()
+    }
+}
+
+/**
+ * Debounce a QUIT received by the IrcHandler to prevent net-splits from spamming leave events
+ * into a room when incremental membership syncing is enabled.
+ * @param {Request} req The metadata request.
+ * @param {IrcServer} server The sending IRC server.
+ * @param {string} matrixUser The virtual user of the user that sent QUIT.
+ * @return {Promise} which resolves to true if a leave should be sent, false otherwise.
+ */
+QuitDebouncer.prototype.debounceQuit = Promise.coroutine(function*(req, server, matrixUser, nick) {
+    // Maintain the last windowMs worth of timestamps corresponding with calls to this function.
+    let now = Date.now();
+    let windowMs = 1000;// Window starts 1s ago
+    let threshold = 5;// Need 5Hz of quits to call net-split
+    this._quitTimesMs[server.domain].push(now);
+
+    // Filter out timestamps from more than 1s ago
+    this._quitTimesMs[server.domain] = this._quitTimesMs[server.domain].filter(
+        (t) => t > now - windowMs
+    );
+
+    // Wait for a short time to allow other potential splitters to send QUITs
+    yield Promise.delay(100);
+
+    let isSplitOccuring = this._quitTimesMs[server.domain].length > threshold;
+
+    if (this._debouncerForServer[server.domain] === undefined) {
+        this._debouncerForServer[server.domain] = {};
+    }
+
+    // TODO: This should be replaced with "disconnected" as per matrix-appservice-irc#222
+    let presence = "offline";
+    try {
+        yield this.ircBridge.getAppServiceBridge().getIntent(
+            matrixUser.getId()
+        ).client.setPresence(presence);
```

Comment:
Presence is per-user, so if they quit, don't we want to set presence anyway? Or should it be 'discconected' for netsplits and 'offline' for a legitimate quit?